File: dbtable.js
/*
	javascript stuff for dbtable application
*/
$(document).ready(function () {
	// refresh dbtable
	$("#ref").click(function () {
		$.ajax({
			url: path + "dbtable",
			data: { 
				table: "pdata",
				idFilt: idF,
				eastingFilt: eastingF,
				northingFilt: northingF,
				elevFilt: elevF,
				d1Filt: d1F,
				d2Filt: d2F 
			}
		}).done(
			function (data) {
				// Parse the JSON data returned.
				var refrData = JSON.parse(data);

				$("#dbtableCont").html(refrData.html);
				if (dialog.mode == "filt") {
					alert("Database filtered. " + refrData.rowcount + " row(s) returned.")
					dialog.mode = "";
				} else {
					alert("Database refreshed!");					
				};
			});
		
		//Change the color of the filter button if the database is filtered.
		if (idF + eastingF + northingF + elevF + d1F + d2F == "") {
			$("#filt").css('background-color', '#117864');
		} else {
			$("#filt").css('background-color', 'red');
		};
	});

	// check all checkboxes with the checkbox in the header
	var checkAll = $("input:checkbox[name='checkAll']");
	checkAll.click(function(){
		if (checkAll.checked){
			checkAll.checked = false
			$("input:checkbox").each (function() { this.checked = false})}
		else {
			checkAll.checked = true
			$("input:checkbox").each (function() { this.checked = true})}
	});


	// delete marked records
	$("#del").click(function () {
		var s = "";
		$(":input:checkbox:checked").
			each(function (ind, val) {
				s += val.name + ";";
			});
		
		//alert(s);
		if (s.length) {
			$.ajax({
				url: path + "dbdel",
				data: { table: "pdata", ids: s }
			}).done(
				function (data) {
					alert(data);
					$("#ref").trigger("click");
				});
		} else {
			alert("No row(s) selected")
		}
	});

	//datepicker
	$(function () {
		$("#d1").datepicker({ dateFormat: 'yy-mm-dd' });
	});

	var selectedId;
	var selectedDate;
	var selectedTime;

	// Initialize filter dialog box variables.
	var idF = "";
	var eastingF = "";
	var northingF = "";
	var elevF = "";
	var d1F = "";
	var d2F = "";

	// Initialize select dialog box variables.
	var idS = "";
	var eastingS = "";
	var northingS = "";
	var elevS = "";
	var d1S = "";
	var d2S = "";
	
	function dialogFill() {
		// By default iput is not allowed for the datepicker.
		$("#d1").datepicker('option', 'constrainInput', true);

		// Selected checkboxes.
		s = $(":input:checkbox:checked");
		
		// Reset the dialog box text.
		$("#filtMessage").html("");

		// Reset the dialog box fields except if filter is selected.
		if (dialog.mode != "filt" || dialog.mode != "sel") {
			$("#diaFields")[0].reset();
		};


		// Filter mode for dialog box.
		if (dialog.mode == "filt") {
			// Allow user input for the date.
			$("#d1").datepicker('option', 'constrainInput', false);

			// Getting the previous filter entries.
			$("#id").val(idF);
			$("#easting").val(eastingF);
			$("#northing").val(northingF);
			$("#elev").val(elevF);
			$("#d1").val(d1F);
			$("#d2").val(d2F);

			// Show filtering options to the user.
			$("#filtMessage").html("<p style='margin-bottom:8px'>RegEx can be used to filter the point ID's.</p>" +
								   "<p style='margin-top:0;margin-bottom:8px'>Valid operators for the easting, northing, elevation, date and hour filtering: " +
								   "<, >, =, <=, >=, and, or, between x and y.</p>" +
								   "<p style='margin-top:0'>Example: < 2017-02-02 and >= 2017-01-03</p>")
		};
		
		// Select mode for dialog box.
		if (dialog.mode == "sel") {
			// Allow user input for the date.
			$("#d1").datepicker('option', 'constrainInput', false);

			// Checking if any checkboxes are thicked and alert the user
			if (s.length >= 1){
				alert("You have already selected row(s)! Clicking 'Save' will delete the current selection!")
			}
			
			// Getting the previous select entries.
			$("#id").val(idS);
			$("#easting").val(eastingS);
			$("#northing").val(northingS);
			$("#elev").val(elevS);
			$("#d1").val(d1S);
			$("#d2").val(d2S);
		
			// Show selecting options to the user.
			$("#filtMessage").html("<p style='margin-bottom:8px'>RegEx can be used to select the point ID's.</p>" +
								   "<p style='margin-top:0;margin-bottom:8px'>Valid operators for the easting, northing, elevation, date and hour selecting: " +
								   "<, >, =, <=, >=, and, or, between x and y.</p>" +
								   "<p style='margin-top:0'>Example: < 2017-02-02 and >= 2017-01-03</p>")
		};
		
		// Update mode for dialog box.
		if (dialog.mode == "upd") {
			var s = $(":input:checkbox:checked");
			if (s.length == 1) {
				var w = s[0].name;	// point id and date time
				var w1 = w.split('|');
				var selectedId = w1[0];
				var w2 = w1[1].split(' ');
				selectedDate = w2[0];
				selectedTime = w2[1];

				$.ajax({
					url: path + "test",
					data: { table: "pdata", id: selectedId,
						d: w1[1]}
				}).done(
					function (data) {
						var res = JSON.parse(data);
						$("#id").val(res[0]);
						$("#easting").val(res[1] == 'None' ? '' : res[1]);
						$("#northing").val(res[2] == 'None' ? '' : res[2]);
						$("#elev").val(res[3] == 'None' ? '' : res[3]);
						w = res[4].split(' ');
						$("#d1").val(w[0]);
						$("#d2").val(w[1]);
					});
			} else if (s.length > 1) {
				alert("Too many rows selected")
				dialog.dialog("close");
			} else {
				alert("Select a row!")
				dialog.dialog("close");
			}
		};

		return[selectedDate, selectedId, selectedTime];
	
	};

	// edit form 
	function save() {
		// Filter option for the dialog box.
		if (dialog.mode == "filt") {
			// Get the values from the filter dialog box.
			idF = $("#id").val();
			eastingF = $("#easting").val();
			northingF = $("#northing").val();
			elevF = $("#elev").val();
			d1F = $("#d1").val();
			d2F = $("#d2").val();

			// Validate the field values.
			// Easting, Northing and Elevation validation:
			if (coordValid(eastingF) == 0) {
				alert("Easting filter invalid!");
				return;
			};

			if (coordValid(northingF) == 0) {
				alert("Northing filter invalid!");
				return;
			};

			if (coordValid(elevF) == 0) {
				alert("Elevation filter invalid!");
				return;
			};

			// D1 validation (year-month-day):
			if (dateValid(d1F) == 0) {
				alert("Date filter invalid!")
				return;
			};

			//D2 validation (hour-minute-second):
			if (hmsValid(d2F) == 0) {
				alert("Time filter invalid!");
				return;
			};

			// Trigger the refresh button with the filter settings.
			$("#ref").trigger("click");
		};

		// Select option for dialog box
		if (dialog.mode == "sel") {
			// uncheck selected checkboxes
			$("input:checkbox:checked").each (function() { this.checked = false});

			// Get the values from the select dialog box.
			idS = $("#id").val();
			eastingS = $("#easting").val();
			northingS = $("#northing").val();
			elevS = $("#elev").val();
			d1S = $("#d1").val();
			d2S = $("#d2").val();

			
		// Validate the field values.
			// Easting, Northing and Elevation validation:
			if (coordValid(eastingS) == 0) {
				alert("Easting select invalid!");
				return;
			};

			if (coordValid(northingS) == 0) {
				alert("Northing select invalid!");
				return;
			};

			if (coordValid(elevS) == 0) {
				alert("Elevation select invalid!");
				return;
			};

			// D1 validation (year-month-day):
			if (dateValid(d1S) == 0) {
				alert("Date select invalid!")
				return;
			};

			//D2 validation (hour-minute-second):
			if (hmsValid(d2S) == 0) {
				alert("Time select invalid!");
				return;
			};


			// if no conditions are given, uncheck all the checkboxes
			if (idS + eastingS + northingS + elevS + d1S + d2S == ""){
				$("input:checkbox:checked").each (function() { this.checked = false})
				dialog.dialog("close")
				return
			};


			$.ajax({
				url: path + "dbtable",
				data: { 
					table: "pdata",
					idFilt: idS,
					eastingFilt: eastingS,
					northingFilt: northingS,
					elevFilt: elevS,
					d1Filt: d1S,
					d2Filt: d2S,
					sel: 1
				}

			// Check the selected checkboxes.
			}).done( function (data) {				
				selData = JSON.parse(data);				
				selData.forEach( function(val) {
					$("input:checkbox[name='" + val + "']")[0].checked = true;			
				})
				alert("Selection is done!")
				dialog.dialog("close");
			})
		};


		// Insert and update
		var loc = { table: "pdata" };
		var fields = ["id", "easting", "northing"];
		var opts = ["elev", "d1", "d2"];
		if (dialog.mode == "ins" || dialog.mode == "upd") {
			for (i = 0; i < fields.length; i++) {
				loc[fields[i]] = $("#" + fields[i]).val().trim();
				if (loc[fields[i]].length == 0) {
					alert(fields[i] + " is empty");
					return loc;
				}
			}
			for (i = 0; i < opts.length; i++) {
				var w = $("#" + opts[i]).val().trim();
				if (w.length == 0) {
					if (dialog.mode == "upd") {
						alert(opts[i] + " is empty");
						return;
					} else {
						w = 'NULL'
					}
				} else {
					loc[opts[i]] = w;
				}
			}

			// Validate fields
			var coordRegex = /^\-?((\d+\.?\d*)|(\.\d+))$/;
			var dateRegex = /^\d{4}\-(0[1-9]|1[0-2])\-(0[1-9]|(1|2)[0-9]|3[0-1])$/;
			var hmsRegex = /^(((0|1)[0-9])|2[0-3])\:([0-5][0-9])\:([0-5][0-9])$/;

			// Easting
			if (coordRegex.test($("#easting").val()) == 0) {
				alert("Invalid easting value!");
				return;
			};

			// Northing
			if (coordRegex.test($("#northing").val()) == 0) {
				alert("Invalid northing value!");
				return;
			};

			// Elevation
			if (coordRegex.test($("#elev").val()) == 0) {
				alert("Invalid elevation value!");
				return;
			};

			// Date
			if (dateRegex.test($("#d1").val()) == 0) {
				alert("Invalid date value!");
				return;
			};

			// Time
			if (hmsRegex.test($("#d2").val()) == 0) {
				alert("Invalid time value!");
				return;
			};

			loc["d"] = loc["d1"] + " " + loc["d2"]
			//del loc["d1"]
			if (dialog.mode == "ins") {
				$.ajax({
					url: path + "dbins",
					data: loc
				}).done(
					function (data) {
						alert(data);
						$("#ref").trigger("click");
					});
			};
		};

			if (dialog.mode == "upd") {
				$.ajax({
					url: path + "dbupd",
					data: loc
				}).done(
					function (data) {
						alert(data);
						$("#ref").trigger("click");
					});
			};

		dialog.dialog("close");
	};

	// Validity checker for the coordinate fields in the filter/select dialog.
	function coordValid(coord) {
		var matches = /^(<|>|=|<=|>=|<>|!=){1}\ ?\-?((\d+\.?\d*)|(\.\d+))$/.test(coord) +
				  /^(<|>|=|<=|>=|<>|!=){1}\ ?\-?((\d+\.?\d*)|(\.\d+))\ (and|or)\ (<|>|=|<=|>=|<>|!=){1}\ ?\-?((\d+\.?\d*)|(\.\d+))$/.test(coord) +
				  /^between\ ?\-?((\d+\.?\d*)|(\.\d+))\ (and)\ ?\-?((\d+\.?\d*)|(\.\d+))$/.test(coord) +
				  /^$/.test(coord);

		if (matches == 0) {
			return(0);
		} else {
			return(1);
		};
	};

	// Validity checker for the date field in the filter/select dialog.
	function dateValid(d1) {
		var matches = /^(<|>|=|<=|>=|<>|!=){1}\ ?\d{4}\-(0[1-9]|1[0-2])\-(0[1-9]|(1|2)[0-9]|3[0-1])$/.test(d1) +
				  /^(<|>|=|<=|>=|<>|!=){1}\ ?\d{4}\-(0[1-9]|1[0-2])\-(0[1-9]|(1|2)[0-9]|3[0-1])\ (and|or)\ (<|>|=|<=|>=|<>|!=){1}\ ?\d{4}\-(0[1-9]|1[0-2])\-(0[1-9]|(1|2)[0-9]|3[0-1])$/.test(d1) +
				  /^between\ \d{4}\-(0[1-9]|1[0-2])\-(0[1-9]|(1|2)[0-9]|3[0-1])\ and\ \d{4}\-(0[1-9]|1[0-2])\-(0[1-9]|(1|2)[0-9]|3[0-1])$/.test(d1) +
				  /^$/.test(d1);
	
		if (matches == 0) {
			return(0);
		} else {
			return(1);
		};
	};

	// Validity checker for the time field in the filter/select dialog.
	function hmsValid(d2) {
		var matches = /^(<|>|=|<=|>=|<>|!=){1}\ ?(((0|1)[0-9])|2[0-3])\:([0-5][0-9])\:([0-5][0-9])$/.test(d2) +
				  /^(<|>|=|<=|>=|<>|!=){1}\ ?(((0|1)[0-9])|2[0-3])\:([0-5][0-9])\:([0-5][0-9])\ (and|or)\ (<|>|=|<=|>=|<>|!=){1}\ ?(((0|1)[0-9])|2[0-3])\:([0-5][0-9])\:([0-5][0-9])$/.test(d2) +
				  /^between\ (((0|1)[0-9])|2[0-3])\:([0-5][0-9])\:([0-5][0-9])\ and\ (((0|1)[0-9])|2[0-3])\:([0-5][0-9])\:([0-5][0-9])$/.test(d2) +
				  /^$/.test(d2);

		if (matches == 0) {
			return(0);
		} else {
			return(1);
		};
	};

	dialog = $("#dia").dialog({
		autoOpen: false, modal: true,
		buttons: {
			Save: save,
			Cancel: function () { dialog.mode = ""; dialog.dialog("close"); }
		}
	});

	$("#dia").dialog({ open: dialogFill });

	// Open the data insert dialog if the "Insert" button is pressed.
	$("#ins").click(function () {
		dialog.mode = "ins";
		dialog.dialog("open");
	});

	// Open the data update dialog if the "Update" button is pressed.
	$("#upd").click(function () {
		dialog.mode = "upd";
		dialog.dialog("open");
	});

	// Open the data selection dialog if the "Select" button is pressed.
	$("#sel").click(function () {
		dialog.mode = "sel";
		dialog.dialog("open");
	});

	// Open the data filter dialog if the "Filter" button is pressed.
	$("#filt").click(function () {
		dialog.mode = "filt";
		dialog.dialog("open");
	});

});

File: dbtable.py
"""
    Python module to handle temporal point data
"""
import logging
import re
import datetime
import psycopg2
import json
import config     # some configuration settings (database, js)

def index(req, table="pdata", order="id"):
    """ creates main html page
        :param req: request data (not used)
        :param table: table name in database
        :param order: sort order of table data
        :returns: html string
    """
    res = """<!doctype html>
        <html>
        <head>
            <meta charset="UTF-8">
            <title> Table list </title>
            <link rel="stylesheet" href="http://code.jquery.com/ui/1.12.1/themes/base/jquery-ui.css">
            <link href="{2}" rel="stylesheet"  type="text/css"/>
            <script type="text/javascript" src="http://code.jquery.com/jquery-3.2.1.min.js" >
            </script>
            <script type="text/javascript" src="http://code.jquery.com/ui/1.12.1/jquery-ui.min.js">
            </script>
            <script type="text/javascript">path="{0}";</script>
            <script type="text/javascript" src="{1}">
            </script>
        </head>
        <body>
            <div id="dia" title="Hello">
                <form>
                    <fieldset>
                    <label for="id">Point ID</label><br />
                    <input type="text" name="id" id="id" size="15"><br />
                    <label for="easting">Easting</label><br />
                    <input type="text" name="easting" id="easting" size="15"><br />
                    <label for="northing">Northing</label><br />
                    <input type="text" name="northing" id="northing" size="15"><br />
                    <label for="elev">Elevation</label><br />
                    <input type="text" name="elev" id="elev" size="15"><br />
                    <label for="d1">Date</label><br />
                    <input type="text" name="d1" id="d1" size="15"><br />
                    <label for="d2">Time</label><br />
                    <input type="text" name="d2" id="d2" size="15">
                    </fieldset>
                </form>
            </div>

            <p class="buttonclass"><input type="button" name="ref" value="Refresh" id="ref">&nbsp;
               <input type="button" name="upd" value="Update" id="upd">&nbsp;
               <input type="button" name="del" value="Delete" id="del">&nbsp;
               <input type="button" name="ins" value="Insert" id="ins">&nbsp;
	           <input type="button" name="sel" value="Select" id="sel">&nbsp;
               <input type="button" name="filt" value="Filter" id="filt">&nbsp;
            </p><div id="dbtable">
        """.format(config.path, config.js, config.css)
    return res + dbtable(req, table, order) + "</div></body></html>"

def dbtable(req, table, order="id", idFilt='', eastingFilt='', northingFilt='', elevFilt='', 
            d1Filt = '', d2Filt = ''):
    """ returns html table of point data from database

        :param req: request data (not used)
        :param table: table name in database
        :param order: sort order of table data
        :param idFilt: filter to query the point id's
        :param eastingFilt: filter for the easting coords
        :param northingFilt: filter for the northing coords
        :param elevFilt: filter for the elevations
        :param d1Filt: first part (year-month-day) of the date filter
        :param d2Filt: second part (hour-minute_second) of the date filter 
        :returns: html string
    """
    logging.basicConfig(format=config.log_format, filename=config.log)
    conn = psycopg2.connect(database=config.database)
    if not conn:
        msg = "Cannot connect to database :("
        logging.error(msg)
        return msg

    sql = "select * from {0} ".format(table)

    query = ""
    # Filtering the database.
    if idFilt != "" or eastingFilt != "" or northingFilt != "" or elevFilt != "" or d1Filt != "" or d2Filt != "":
        query += "where "

    # Point ID filter
    if idFilt != "":
        query += "id ~ '{0}' and ".format(idFilt)

    # Easting filter
    if eastingFilt != "":
        query += "easting {0} and ".format(eastingFilt)

    # Northing filter
    if northingFilt != "":
        query += "northing {0} and ".format(northingFilt)

    # Elevation filter
    if elevFilt != "":
        query += "elev {0} and ".format(elevFilt)

    # Year-Month-Day filter
    if d1Filt != "":
        d1FiltMod = re.subn(r'(\d{4}\-\d{2}\-\d{2})', r"'\1'", d1Filt)[0]
        query += "to_char(d, 'YYYY-MM-DD') {0} and ".format(d1FiltMod)

    # Hour-Minute-Second filter
    if d2Filt != "":
        d2FiltMod = re.subn(r'(\d{2}\:\d{2}\:\d{2})', r"'\1'", d2Filt)[0]
        query += "to_char(d, 'HH24:MI:SS') {0}".format(d2FiltMod)

    # Strip the ending 'and' of the SQL and close it
    query = query.strip(" and ")
    sql += query

    if order is not None and len(order.strip()):
        sql += " order by {0}".format(order)
    logging.debug(sql)
    cur = conn.cursor()
    cur.execute(sql)
    # hibakezeles!!!!

    res = """
             <table border="1">
             <tr><th>&nbsp;</th><th>ID</th><th>Easting</th><th>Northing</th><th>Elevation</th><th>Time</th></tr>
        """
    for row in cur:
        res += """<tr><td><input type="checkbox" name="{0}|{3}"></td>
               <td>{0}</td><td>{1}</td>
               <td>{2}</td>
               <td>{3}</td>
               <td>{4}</td></tr>""".format(row[0], row[1], row[2], row[3], row[4])
    res += "</table>"
    cur.close()
    return res

def dbdel(req, table, ids):
    """ Delete records from table, ids contains point id and date/time in the
        form point_id|datetime;point_id|datetime;...
        :param req: request object from apache
        :param table: table name
        :param ids: keys to records to delete
        :return: number of deleted rows
    """
    logging.basicConfig(format=config.log_format, filename=config.log)
    conn = psycopg2.connect(database=config.database)
    if not conn:
        msg = "Cannot connect to database :("
        logging.error(msg)
        return msg
    sql = "delete from {0} where (id, d) in (".format(table)
    keys = ids.strip(";").split(";")
    for key in keys:
        i, d = key.split("|")
        sql += "('{0}', '{1}'),".format(i, d)
    sql = sql.strip(',')
    sql += ")"
    logging.debug(sql)
    cur = conn.cursor()
    cur.execute(sql)
    msg = "{0} lines deleted".format(cur.rowcount)
    logging.info(msg)
    conn.commit()
    cur.close()
    return msg

def test(req, table, id, d):
    logging.basicConfig(format=config.log_format, filename=config.log)
    conn = psycopg2.connect(database=config.database)
    if not conn:
        msg = "Cannot connect to database :("
        logging.error(msg)
        return msg

    sql = "select id,easting,northing,elev,d from {0} where id='{1}' and d = '{2}'".format(table, id, d)
    logging.debug(sql)
    cur = conn.cursor()
    cur.execute(sql)
    if cur.rowcount == 1:
        result = json.dumps(['{}'.format(x) for x in cur.fetchone()])
    else:
        result = []

    # hibakezeles!!!
    cur.close()
    return result

def dbins(req, table, id, easting, northing, elev=None, d=None):
    """ Insert new record into table
        :param req: request object from apache
        :param table: table name
        :param id: point id
        :param easting: easting coordinate
        :param northing: northing coordinate
        :param elev: elevation
        :param d: data and time
        :return: number of deleted rows
    """
    logging.basicConfig(format=config.log_format, filename=config.log)
    conn = psycopg2.connect(database=config.database)
    if not conn:
        msg = "Cannot connect to database :("
        logging.error(msg)
        return msg
    if elev is None:
        elev = 'NULL'
    if d is None:
        d = datetime.datetime.now().isoformat(" ")
    sql = """insert into {0} (id, easting, northing, elev, d)
        values ('{1}', {2}, {3}, {4}, '{5}')
        """.format(table, id, easting, northing, elev, d)
    logging.debug(sql)
    cur = conn.cursor()
    cur.execute(sql)
    msg = "{0} lines inserted".format(cur.rowcount)
    logging.info(msg)
    conn.commit()
    cur.close()
    return msg


def dbupd(req, table, id, easting, northing, elev=None, d=None):
    """ Insert new record into table
        :param req: request object from apache
        :param table: table name
        :param id: point id
        :param easting: easting coordinate
        :param northing: northing coordinate
        :param elev: elevation
        :param d: data and time
        :return: number of deleted rows
    """
    logging.basicConfig(format=config.log_format, filename=config.log)
    conn = psycopg2.connect(database=config.database)
    if not conn:
        msg = "Cannot connect to database :("
        logging.error(msg)
        return msg
    if elev is None:
        elev = 'NULL'
    if d is None:
        d = datetime.datetime.now().isoformat(" ")
    sql = """UPDATE {0} 
            SET id='{1}', easting='{2}', northing='{3}', elev='{4}', d='{5}'
            WHERE id='{1}';
        """.format(table, id, easting, northing, elev, d)
    logging.debug(sql)
    cur = conn.cursor()
    cur.execute(sql)
    msg = "Line: '{0}' updated".format(id)
    logging.info(msg)
    conn.commit()
    cur.close()
    return msg